packages/components/src/color-picker/index.js

```diff
@@ -43,30 +43,81 @@ import Alpha from './alpha';
 import Hue from './hue';
 import Inputs from './inputs';
 import Saturation from './saturation';
-import { colorToState, simpleCheckForValidColor } from './utils';
+import { colorToState, simpleCheckForValidColor, isValidHex } from './utils';
+
+const toLowerCase = ( value ) => String( value ).toLowerCase();
 
 export default class ColorPicker extends Component {
 	constructor( { color = '0071a1' } ) {
 		super( ...arguments );
-		this.state = colorToState( color );
+		const colors = colorToState( color );
+		this.state = {
+			...colors,
+			draftHex: toLowerCase( colors.hex ),
+			draftRgb: colors.rgb,
+			draftHsl: colors.hsl,
+		};
 		this.handleChange = this.handleChange.bind( this );
+		this.handleInputChange = this.handleInputChange.bind( this );
 	}
 
 	handleChange( data ) {
 		const { oldHue, onChangeComplete = noop } = this.props;
-		const isValidColor = simpleCheckForValidColor( data );
-		if ( isValidColor ) {
+		const isValid = data.hex ? isValidHex( data.hex ) : simpleCheckForValidColor( data );
+		if ( isValid ) {
 			const colors = colorToState( data, data.h || oldHue );
-			this.setState(
-				colors,
-				debounce( partial( onChangeComplete, colors ), 100 )
+			this.setState( {
+				...colors,
+				draftHex: toLowerCase( colors.hex ),
+				draftHsl: colors.hsl,
+				draftRgb: colors.rgb,
+			},
+			debounce( partial( onChangeComplete, colors ), 100 )
 			);
 		}
 	}
 
+	handleInputChange( data ) {
+		if ( data.state === 'commit' ) {
+			this.handleChange( data );
+		} else if ( data.state === 'draft' ) {
```

Comment:
By lifting the draft cache to the `ColorPicker` and updating state we are forcing the other views to re-render, which may cause performance issues. We should test and avoid re-renderings if necessary.

Reply:
Yes, we're re-rendering saturation and hue every time an input changes. Can we keep the draft in the Inputs state until we're ready to commit, and then call this.props.onChange?

It seems to me that `ColorPicker` knows too much about the child components. If the child components only took a hex value color, they could process it into whatever values they needed. For example, `Inputs` could split it into rgba, and when the internal draft state was committed by calling `this.props.onChange`, the `ColorPicker` state would change, which would send the new `color` to all the child components, and they would update their internal state with `getDerivedStateFromProps`.

Reply:
I wish we could. This PR was so involved because we cannot longer do that. I've described why we can't do that in this issue description. Does that help?

In 520c3e1 and be0cc4f I've prevented the other components from re-render, which alleviates my original concerns.

Reply:
Good that the re-render is fixed :)

I'm sorry, I totally misunderstood the description. You're totally right. Wow, that seems like quite a step backwards, to me. Ok, I understand now, thanks :)